src/dyn/Dynamics.cpp

```diff
@@ -1562,8 +1573,10 @@ void compute_dynamics(dyn_variables& dyn_var, bp::dict dyn_params,
     if(prms.thermally_corrected_nbra==1 && prms.tcnbra_do_nac_scaling==1){  remove_thermal_correction(dyn_var, ham, prms);  }
     
     // Update vib Hamiltonian to reflect the change of the momentum
-    update_Hamiltonian_variables(prms, dyn_var, ham, ham_aux, py_funct, params, 1); 
-        
+    if(prms.use_qtsh==0){

```

Comment:
i don't agree with this conditional call (nested inside if block) of this function - basically, is should be always called but would do nothing for QTSH. You are right - this could be more efficient your way, but adds the complexity of the code. In this case I'd sacrifice the efficiency for the sake of clarity and simplicity